src/box/replication.cc

```diff
@@ -653,6 +653,10 @@ replicaset_update(struct applier **appliers, int count, bool keep_connect)
 	replicaset.applier.connected = 0;
 	replicaset.applier.loading = 0;
 	replicaset.applier.synced = 0;
```

Comment:
Why do we even nullify those counters? Wouldn't it be more natural to let them be updated via the normal flow? Not break it by this nullification? So we keep the counters unchanged, destroy the appliers, they will themselves properly decrement all the counters, and all will be fine. Then you wouldn't need to rely on which functions yield and which not. For example, currently your fix won't work anymore if `replica_clear_applier` or `replica_delete` would start yielding.

Reply:
The short answer is that it's too difficult to implement that fix for such a rare bug. Here's the explanation.

Every applier has a list of trigger, the second of which is the one, we're interested in. It's set as soon as we created replica and assign applier to it. Applier cannot belong to several replicas at the same time.

On configuration the list of appliers is created in `box_sync_replication`. Appliers are started, and only after that they go down, to `replicaset_update`, where replicas are created, applier are assigned to them. By this time applier can have any state already. So, triggers didn't update state machine. 

The fix could be to update state machine right from applier, but not from triggers. But it would break consistency of the code.